src/GridFS/GridFSAdapter.php

```diff
@@ -0,0 +1,404 @@
+    {
+        if (str_ends_with($path, '/')) {
+            throw UnableToRetrieveMetadata::fileSize($path, 'file path cannot end with a slash');
+        }
+
+        $file = $this->findFile($path);
+
+        if ($file === null) {
+            throw UnableToRetrieveMetadata::fileSize($path, 'file does not exist');
+        }
+
+        return new FileAttributes($path, $file['length']);
+    }
+
+    public function listContents(string $path, bool $deep): iterable
+    {
+        $path = $this->prefixer->prefixDirectoryPath($path);
+
+        $pathdeep = 0;
+        $pipeline = [];
+        if ($path !== '') {
+            $pathdeep = substr_count($path, '/');
+            // Exclude files that does not start with the expected path
+            $pipeline[] = ['$match' => ['filename' => ['$regex' => new Regex('^' . preg_quote($path))]]];
+        }
+        // Get the last revision of each file
+        $pipeline[] = ['$sort' => ['filename' => 1, 'uploadDate' => -1]];
+
+        if ($deep === false) {
+            $pipeline[] = ['$addFields' => ['splitpath' => ['$split' => ['$filename', '/']]]];
+            $pipeline[] = ['$group' => [
+                // The same name could be used as a filename and as part of the path of other files
+                '_id' => [
+                    'basename' => ['$arrayElemAt' => ['$splitpath', $pathdeep]],
+                    'isDir' => ['$ne' => [['$size' => '$splitpath'], $pathdeep + 1]],
+                ],
+                // Get the metadata of the last revision of each file
+                'file' => ['$first' => '$$ROOT'],
+                // The "lastModified" date is the date of the last uploaded file in the directory
+                'uploadDate' => ['$max' => '$uploadDate'],
+            ]];
+
+            $files = $this->bucket->getFilesCollection()->aggregate(
+                $pipeline,
+                ['typeMap' => ['root' => 'array', 'document' => 'array', 'array' => 'array']]
+            );
+
+            foreach ($files as $file) {
+                if ($file['_id']['isDir']) {
+                    yield new DirectoryAttributes(
+                        $this->prefixer->stripDirectoryPrefix($path . $file['_id']['basename']),
+                        null,
+                        $file['uploadDate']->toDateTime()->getTimestamp(),
+                    );
+                } else {
+                    $file = $file['file'];
+                    yield new FileAttributes(
+                        $this->prefixer->stripPrefix($file['filename']),
+                        $file['length'],
+                        $file['metadata'][self::METADATA_VISIBILITY] ?? null,
+                        $file['uploadDate']->toDateTime()->getTimestamp(),
+                        $file['metadata'][self::METADATA_MIMETYPE] ?? null,
+                        $file,
+                    );
+                }
+            }
+        } else {
+            // Get the last revision of each file name
```

Comment:
```suggestion
            // Get the metadata of the last revision of each file
```

For consistency with the earlier pipeline construction. I assume these should be consistent.